pkg/kubelet/kuberuntime/kuberuntime_image.go

```diff
@@ -105,6 +108,16 @@ func (m *kubeGenericRuntimeManager) ListImages(ctx context.Context) ([]kubeconta
 	}
 
 	for _, img := range allImages {
+		// When RuntimeClassInImageCriAPI feature gate is set, it is expected that the necessary
+		// changes needed for KEP 4216 has made it to the container runtime being used as well.
+		// Therefore, validate that image returned from ListImages() call above has a valid
+		// runtimehandler field set; that is, it should not be ""
+		if utilfeature.DefaultFeatureGate.Enabled(features.RuntimeClassInImageCriAPI) {
+			if img.Spec != nil && img.Spec.RuntimeHandler == "" {
```

Comment:
you should be able to show that in the mock test linked above ^

Reply:
@mikebrow added test. Please let me know if this looks good. Thanks.

Reply:
The only intention here is to have a uniform way to identify as an image as its name/ID + runtime class used to pull the image. Does k8s allow us to always have the feature gate to support this feature? If yes, we can enable the feature gate for runtimes that implement this feature. If not, I am not sure how clean the implementation of this feature would be to have some images identified as name and empty "" runtime handler and in some other cases images to be identified as name and a non-empty runtime handler. 
For example is this behavior acceptable - 
 1. Lets say with CRI-O runtime, some user specifies a runtime handler in the deployment yaml, then the EnsureImageExists() call is going to pass the runtime handler down to the CRI-O container runtime which may or may not choose to use this field. Likewise, when kubelet GC calls ListImages() from the container runtime, then CRI-O responds but does  not set the runtime handler field and GC continues to function as it is today.
 2. Lets say the container runtime is containerd which implements image pull per runtime class changes on its side completely (that is, an image on containerd will be identified as a tuple of imageName/imageId, runtimeHandler and if an empty runtimeHandler is passed down to it via CRI, then it uses the default runtime handler for that platform.
 Likewise, when kubelet GC calls ListImages() from containerd, containerd returns the list of images with the imageName and the runtime handler used to pull that image and then kubelet  GC uses both these fields to decide which image it wants to delete.

If the above behavior is acceptable from k8s side and seems like a decent solution to have given its flexibility, we could do as @haircommander is suggesting. Thoughts @mikebrow @marosset @jsturtevant @aravindhp ?

Reply:
Had a discussion with @mikebrow over a call now. Here's what might be a better approach:

Had a discussion with @mikebrow over a call now. Here's what might be a better approach:

1. On image pull and garbage collection:

- if runtime handler is specified in the yaml, EnsureImageExists() will pass on the imageName and runtimeHandler field to the container runtime and the container runtime
can choose to use this field or not.
- In kubelet image GC, the following is an important data structure used:

type realImageGCManager struct {
	// Container runtime
	runtime container.Runtime

	// Records of images and their use.
	imageRecords     map[string]*imageRecord  >>>>>>>>>>>>>>> map of ImageRecord struct that is indexed by ImageID
	imageRecordsLock sync.Mutex

	// The image garbage collection policy in use.
	policy ImageGCPolicy

	// statsProvider provides stats used during image garbage collection.
	statsProvider StatsProvider

	// Recorder for Kubernetes events.
	recorder record.EventRecorder

	// Reference to this node.
	nodeRef *v1.ObjectReference

	// Track initialization
	initialized bool

	// imageCache is the cache of latest image list.
	imageCache imageCache

	// tracer for recording spans
	tracer trace.Tracer
}


the detectImages() function creates new imageRecord entries for new images after calling container runtime's ListImages() call. The index to realImageGCManager.imageRecords[] should now 
be a tuple of  (imageID, runtimeHandler) instead of just the imageID.

If the container runtime implements KEP 4216, then ListImages() call will return a list of imageID, runtimeHandler. If runtimehandler == "" , it would just mean that the container runtime does not implement KEP 4216. If runtime != "", then great the container runtime implements KEP 4216 like containerd.


2. What if the user changes the definition of the runtime handler on container runtime side like containerd. 

Example:

(i) a runtime class could have orginially been

          [plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runhcs-wcow-hyperv.guest_platform]
            Architecture = "amd64"
            OS = "windows"
            OSFeatures = []
            OSVersion = "10.0.20348"
            Variant = ""


(ii) Then it was changed to the following:

          [plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runhcs-wcow-process.guest_platform]
            Architecture = "amd64"
            OS = "windows"
            OSFeatures = []
            OSVersion = "10.0.17763"
            Variant = ""

When runtimeclass information is changed, this would need a restart of containerd but the old images pulled with (i) will still be loaded back on containerd restart and kubelet will think the image exists but the runhcs-wcow-process.guest_platform information has changed. Therefore, we need to have a TODO in containerd code and also k8s documentation for this KEP
that if the runtimeclass's information changes, we need to flush containerd and restart both containerd and kubelet.


cc @jsturtevant @marosset @aravindhp @haircommander

Reply:
> Does k8s allow us to always have the feature gate to support this feature?

The typical pattern is for it to be turned on by default in `Beta` then locked to on once in `GA`.  With the feature gate being dropped after that

I was under the impression that empty string would equate to default runtime handler and there for not throw an error here based on
https://github.com/kubernetes/kubernetes/blob/6fed03ea91627185afd4976474a396f17689f0d7/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto#L787

> Lets say with CRI-O runtime, some user specifies a runtime handler in the deployment yaml, then the EnsureImageExists() call is going to pass the runtime handler down to the CRI-O container runtime which may or may not choose to use this field.

This sounds like it could be an edge case where the pulled image might not be the expected version.  Should we log a warning here in this case?  In this case I think user would experience the same errors they do with out this feature?

Reply:
> If runtimehandler == "" , it would just mean that the container runtime does not implement KEP 4216

You still have the situation described with CRI-O. Is the KEP 4216 something runtimes must implement? if so a failure here is warranted otherwise a warning makes sense to me.

Reply:
> I was under the impression that empty string would equate to default runtime handler and there for not throw an error here based on

>> This is true on the container runtime side when we implement the support for this feature. If runtimehandler returned in ListImages() is "", kubelet is not aware of default runtime

Reply:
> > If runtimehandler == "" , it would just mean that the container runtime does not implement KEP 4216
> 
> You still have the situation described with CRI-O. Is the KEP 4216 something runtimes must implement? if so a failure here is warranted otherwise a warning makes sense to me.

I don't think KEP 4216 should be a must implement for all runtimes. if no runtimeHandler is returned in the list images response from the container runtime, then the index tuple for this image should be (imageId, "") in the imageRecords map of kubelet GC

Reply:
>that if the runtimeclass's information changes, we need to flush containerd and restart both containerd and kubelet.

Will this require the node to be cordoned and drained? Not sure who the "we" here is that will restart both the services.

Reply:
The admin decides if runtime class definition was changed and flushes/remove images from containerd  and restarts. For kubelet a simple service restart should suffice right as it doesn't see to persist the tables that GC looks at. Not sure if node should be cordoned. Should they?

Reply:
> If runtimehandler returned in ListImages() is "", kubelet is not aware of default runtime

We should make this clear in the docs as an expectation.  

> I don't think KEP 4216 should be a must implement for all runtimes.

this should not error here then, I still think there is an edge case where the user specifies the runtimeclass and it is sent to the runtime and it doesn't handle it properly and it should be logged as a warning.

Reply:
@jsturtevant Agreed. I haven't made changes to reflect these in the PR yet as I was looking for feedback/arrive at a consensus for the proposal I made earlier about desired behavior. I'll work on these now and push need changes.

Reply:
>The admin decides if runtime class definition was changed and flushes/remove images from containerd and restarts. For kubelet a simple service restart should suffice right as it doesn't see to persist the tables that GC looks at. Not sure if node should be cordoned. Should they?

It was not immediately obvious to me that admin intervention was required. Not sure if this is a great UX but I will defer to SIG-node here.

Reply:
> > The admin decides if runtime class definition was changed and flushes/remove images from containerd and restarts. For kubelet a simple service restart should suffice right as it doesn't see to persist the tables that GC looks at. Not sure if node should be cordoned. Should they?
> 
> It was not immediately obvious to me that admin intervention was required. Not sure if this is a great UX but I will defer to SIG-node here.

Not sure that there is a better way to do this. If not, the existing images can be allowed to be loaded and used by containerd and eventually starting containers could fail and then user would have to flush containerd of the older images. The latter might be a painful debugging experience. There it might be better to flush containerd and restart just like in toggling the feature gate for this KEP would need manual intervention. 
Alternatively, we could add an option/annotation/config on containerd side which when set in the toml causes containerd to purge and delete the images using the edited runtime class rather than loading them. Any thoughts on this last option @mikebrow @dmcgowan?

Reply:
we had to make a similar concession in the [cgroup driver reported by runtime KEP](https://github.com/kubernetes/enhancements/issues/4033). Basically, it's up to the container runtime/admin to verify this information isn't changing while there are containers/images that already exist. Kubelet has no way to verify it mid run (or it would be inefficent to do so). Explicitly declaring this is a requirement of the CRI is good here, and having a mechanism in the CRI to report errors to the admin would be good to (but out of scope here)

Reply:
this invariant https://github.com/kubernetes/kubernetes/pull/118770/files#diff-7a3c2723dedc6b44cf183fc52f84cc4015ffbd65a998a65c3959a953cb63c945R139-R141 it's declared in the CRI definition

Reply:
> this invariant https://github.com/kubernetes/kubernetes/pull/118770/files#diff-7a3c2723dedc6b44cf183fc52f84cc4015ffbd65a998a65c3959a953cb63c945R139-R141 it's declared in the CRI definition

This is great! 
"The Kubelet will not re-request the RuntimeConfiguration after startup, and CRI implementations should avoid updating them without a full node reboot." 
-- Node reboot might still not fix containerd loading previous images in containerd's store. I discussed this in the containerd community meeting this morning and we are fine with the existing invariant which is to ask admins to purge if runtimeclass name is the same but some of its fields like platform versioning and snapshotters have changed. 

cc @mikebrow @dmcgowan

Reply:
@haircommander @jsturtevant @aravindhp @mikebrow pushed new changes based on the discussion. Updated the tests as necessary. 
Please let me know if it looks ok :) Thanks a lot!